File: javascript/leap/leap.js
var Year = function(year) {
  this.year = year;
};

var LEAP_YEAR = true,
    COMMON_YEAR = false;

var isDivisibleBy = function isDivisibleBy(year, denominator) {
  return year % denominator == 0
}

Year.prototype.isLeap = function() {
  if (!isDivisibleBy(this.year, 4)) {
    return COMMON_YEAR;
  } else if (!isDivisibleBy(this.year, 100)) {
    return LEAP_YEAR;
  } else if (!isDivisibleBy(this.year, 400)) {
    return COMMON_YEAR;
  } else {
    return LEAP_YEAR;
  }
};

module.exports = Year;

File: scala/leap/src/main/scala/leap.scala
class Year(year: Int) {

  implicit class IntHelpers(val i: Int) {
    def isDivisibleBy(factor: Int): Boolean = i % factor == 0
  }

  def isLeap: Boolean = {
    year.isDivisibleBy(400) ||
      (year.isDivisibleBy(4) && !year.isDivisibleBy(100))
  }

}

object Year {
  def apply(year: Int): Year = {
    new Year(year)
  }
}
